GetIntoTeachingApi/Models/BaseModel.cs

```diff
@@ -118,6 +118,14 @@ protected void NotifyPropertyChanged([CallerMemberName] string propertyName = ""
             }
         }
 
+        protected void DeleteLink(ICrmService crm, OrganizationServiceContext context, BaseModel childToRemove, string childToRemovePropertyName)
+        {
+            PropertyInfo property = GetType().GetProperty(childToRemovePropertyName);
+            var attribute = EntityRelationshipAttribute(property);
+            var target = childToRemove.ToEntity(crm, context);
+            crm.DeleteLink(MapToEntity(crm, context), new Relationship(attribute.Name), target, context);

```

Comment:
Re-mapping to an entity here feels off -- I wonder if instead of doing the logic in `ShouldMap` would we be better having another callback that fits in later, something like:

```
var entity = crm.MappableEntity(LogicalName(GetType()), Id, context);
MapFieldAttributesToEntity(entity);
MapRelationshipAttributesToEntity(entity, crm, context);
FinaliseEntity(crm, context);

...

public void FinaliseEntity(ICrmService crm, OrganizationServiceContext context)
{
...
DeleteLink(crm, context, existingEvent.Building, nameof(existingEvent.Building));
...
base.FinaliseEntity(crm, context);
}
```

Then you have access to the `source` already with `this`?

```
crm.DeleteLink(this, new Relationship(attribute.Name), context);
```